File: src/lib/science.ts
import { n } from "./format";
import type { Activity, DayLog, Energy, Pace, Profile, Sex, TimeBudget } from "./types";

/** طاقة تقريبية لكل كغ من نسيج الجسم المفقود (خليط دهون وعضل). */
export const KCAL_PER_KG = 7700;

const ACTIVITY_FACTOR: Record<Activity, number> = {
  sedentary: 1.2,
  light: 1.375,
  moderate: 1.55,
  active: 1.725,
};

/** نسبة النزول الأسبوعي من وزن الجسم. أقصى حد آمن ~1٪ أسبوعياً. */
const PACE_RATE: Record<Pace, number> = {
  gentle: 0.0035,
  steady: 0.006,
  brisk: 0.009,
};

export const CALORIE_FLOOR: Record<Sex, number> = { f: 1200, m: 1500 };

const round = (n: number, step = 1) => Math.round(n / step) * step;

export function bmi(weightKg: number, heightCm: number): number {
  const h = heightCm / 100;
  return weightKg / (h * h);
}

/** معادلة Mifflin-St Jeor — الأدق بين المعادلات الشائعة للبالغين. */
export function bmr(sex: Sex, weightKg: number, heightCm: number, age: number): number {
  const base = 10 * weightKg + 6.25 * heightCm - 5 * age;
  return sex === "m" ? base + 5 : base - 161;
}

export function formulaTdee(p: Pick<Profile, "sex" | "age" | "heightCm" | "activity">, weightKg: number): number {
  return bmr(p.sex, weightKg, p.heightCm, p.age) * ACTIVITY_FACTOR[p.activity];
}

/**
 * وزن مرجعي للبروتين: عند زيادة الوزن نستخدم وزناً معدّلاً
 * (الوزن عند BMI 25 + ربع الفائض) حتى لا يتضخم الهدف بلا داعٍ.
 */
export function referenceWeight(weightKg: number, heightCm: number): number {
  const h = heightCm / 100;
  const at25 = 25 * h * h;
  if (weightKg <= at25) return weightKg;
  return at25 + 0.25 * (weightKg - at25);
}

/** 1.2–1.6 غ/كغ أثناء النزول للحفاظ على العضل؛ نستهدف 1.5 من الوزن المرجعي. */
export function proteinTarget(weightKg: number, heightCm: number): number {
  return round(1.5 * referenceWeight(weightKg, heightCm), 5);
}

export function weeklyLossKg(pace: Pace, weightKg: number): number {
  return PACE_RATE[pace] * weightKg;
}

export interface Targets {
  tdee: number;
  kcal: number;
  protein: number;
  weeklyLossKg: number;
  steps: number;
  water: number;
  floorApplied: boolean;
}

export function computeTargets(profile: Profile, weightKg: number, tdeeOverride?: number): Targets {
  const tdee = tdeeOverride ?? formulaTdee(profile, weightKg);
  const loss = weeklyLossKg(profile.pace, weightKg);
  const deficit = (loss * KCAL_PER_KG) / 7;
  const raw = tdee - deficit;
  const floor = CALORIE_FLOOR[profile.sex];
  const kcal = Math.max(raw, floor);
  const baseSteps = { sedentary: 6000, light: 7000, moderate: 8000, active: 9000 }[profile.activity];
  return {
    tdee: round(tdee, 10),
    kcal: round(kcal, 10),
    protein: proteinTarget(weightKg, profile.heightCm),
    weeklyLossKg: Math.round(loss * 100) / 100,
    steps: baseSteps,
    water: 8,
    floorApplied: raw < floor,
  };
}

/* ------------------------------------------------------------------ */
/* الوزن الاتجاهي: متوسط متحرك أُسّي يخفي تذبذب الماء والملح           */
/* ------------------------------------------------------------------ */

export interface WeightPoint {
  date: string;
  kg: number;
}

export interface TrendPoint extends WeightPoint {
  trend: number;
}

export function daysBetween(a: string, b: string): number {
  const ms = Date.parse(`${b}T00:00:00Z`) - Date.parse(`${a}T00:00:00Z`);
  return Math.round(ms / 86_400_000);
}

/**
 * EMA بمعامل 0.1 يومياً، مع مراعاة الأيام الفارغة:
 * كلما طالت الفجوة زاد وزن القراءة الجديدة.
 */
export function trendWeights(points: WeightPoint[], alpha = 0.1): TrendPoint[] {
  const sorted = [...points].sort((a, b) => a.date.localeCompare(b.date));
  const out: TrendPoint[] = [];
  let trend: number | null = null;
  let last: string | null = null;
  for (const p of sorted) {
    if (trend === null || last === null) {
      trend = p.kg;
    } else {
      const gap = Math.max(1, daysBetween(last, p.date));
      const a = 1 - Math.pow(1 - alpha, gap);
      trend = trend + a * (p.kg - trend);
    }
    last = p.date;
    out.push({ ...p, trend: Math.round(trend * 100) / 100 });
  }
  return out;
}

/* ------------------------------------------------------------------ */
/* الحرق التكيّفي: نتعلم حرقك الحقيقي من أكلك المسجّل وتغيّر وزنك      */
/* ------------------------------------------------------------------ */

export interface AdaptiveResult {
  tdee: number;
  confidence: number;
  loggedDays: number;
  windowDays: number;
}

export function dayIntake(day: DayLog): number {
  return day.meals.reduce((s, m) => s + m.kcal, 0);
}

/**
 * نافذة حتى 21 يوماً: الحرق ≈ متوسط الأكل − (تغيّر الوزن الاتجاهي × 7700 ÷ الأيام).
 * نمزجه مع تقدير المعادلة بحسب كمية البيانات، ونحدّ التغيير بـ ±25٪.
 */
export function adaptiveTdee(profile: Profile, days: DayLog[], today: string): AdaptiveResult | null {
  const window = days
    .filter((d) => daysBetween(d.date, today) >= 0 && daysBetween(d.date, today) < 21)
    .sort((a, b) => a.date.localeCompare(b.date));
  const intakeDays = window.filter((d) => dayIntake(d) >= 600);
  const weights = window.filter((d) => d.weightKg !== null).map((d) => ({ date: d.date, kg: d.weightKg as number }));
  if (intakeDays.length < 7 || weights.length < 3) return null;

  const trend = trendWeights(weights);
  const first = trend[0];
  const lastT = trend[trend.length - 1];
  const span = daysBetween(first.date, lastT.date);
  if (span < 7) return null;

  const avgIntake = intakeDays.reduce((s, d) => s + dayIntake(d), 0) / intakeDays.length;
  const measured = avgIntake - ((lastT.trend - first.trend) * KCAL_PER_KG) / span;
  const formula = formulaTdee(profile, lastT.trend);
  const confidence = Math.min(1, (intakeDays.length / 14) * Math.min(1, span / 14));
  const blended = formula + (measured - formula) * confidence;
  const clamped = Math.min(formula * 1.25, Math.max(formula * 0.75, blended));
  return {
    tdee: round(clamped, 10),
    confidence: Math.round(confidence * 100) / 100,
    loggedDays: intakeDays.length,
    windowDays: span,
  };
}

/* ------------------------------------------------------------------ */
/* الخيط: لا تفوّت مرتين — اليوم الواحد الضائع لا يقطع السلسلة         */
/* ------------------------------------------------------------------ */

export function isCounted(day: DayLog | undefined): boolean {
  if (!day) return false;
  return day.done.length > 0 || day.meals.length > 0 || day.workouts.length > 0 || day.weightKg !== null;
}

export interface ThreadState {
  length: number;
  /** أمس ضاع: اليوم إنقاذ للخيط */
  rescueToday: boolean;
  best: number;
}

export function addDays(date: string, n: number): string {
  const d = new Date(`${date}T00:00:00Z`);
  d.setUTCDate(d.getUTCDate() + n);
  return d.toISOString().slice(0, 10);
}

export function computeThread(days: Record<string, DayLog>, today: string): ThreadState {
  const keys = Object.keys(days).sort();
  if (keys.length === 0) return { length: 0, rescueToday: false, best: 0 };

  let best = 0;
  let run = 0;
  let misses = 0;
  let cursor = keys[0];
  while (cursor <= today) {
    if (isCounted(days[cursor])) {
      run += 1;
      misses = 0;
    } else if (cursor !== today) {
      misses += 1;
      if (misses >= 2) run = 0;
    }
    best = Math.max(best, run);
    cursor = addDays(cursor, 1);
  }
  const yesterday = addDays(today, -1);
  const rescueToday = run > 0 && !isCounted(days[yesterday]) && !isCounted(days[today]) && keys[0] <= yesterday;
  return { length: run, rescueToday, best };
}

/* ------------------------------------------------------------------ */
/* خطة اليوم: تتقلص وتكبر حسب طاقتك ووقتك — لا يوجد يوم صفر          */
/* ------------------------------------------------------------------ */

export interface Mission {
  id: string;
  kind: "move" | "eat" | "restore";
  title: string;
  detail: string;
  routineId?: string;
}

export function dayMissions(energy: Energy, time: TimeBudget, t: Targets, profile: Profile): Mission[] {
  const proteinMeal = Math.round(t.protein / 3 / 5) * 5;
  const move: Mission =
    energy === 1 || time === 2
      ? {
          id: "move",
          kind: "move",
          title: "دقيقتين حركة فقط",
          detail: "٣ تمارين هادئة وأنت بمكانك. الهدف تحافظ على الخيط، مو تتعب.",
          routineId: energy === 1 ? "reset-2" : "wake-2",
        }
      : time === 10
        ? {
            id: "move",
            kind: "move",
            title: energy === 3 ? "١٠ دقائق قوة للجسم كامل" : "١٠ دقائق حركة بدون قفز",
            detail: "تمارين بوزن الجسم تحمي عضلاتك وأنت تنزل وزن.",
            routineId: energy === 3 ? "strength-10" : "low-impact-10",
          }
        : {
            id: "move",
            kind: "move",
            title: energy === 3 ? "٢٠ دقيقة قوة + مشي" : "٢٠ دقيقة مشي خفيف",
            detail: energy === 3 ? "جلسة قوة كاملة؛ أهم استثمار للحفاظ على العضل." : `امشِ بإيقاع مريح. هدف خطواتك ${n(t.steps)}.`,
            routineId: energy === 3 ? "strength-20" : "walk-20",
          };

  const eat: Mission =
    energy === 1
      ? {
          id: "eat",
          kind: "eat",
          title: "ابدأ وجبتك الجاية بالبروتين",
          detail: `حوالي ${n(proteinMeal)} غ (بيض، زبادي، دجاج، تونة). التعب يرفع الجوع — البروتين يهدّيه.`,
        }
      : {
          id: "eat",
          kind: "eat",
          title: "سجّل وجباتك — ولو بجملة",
          detail: `قل لسند "تغديت كبسة دجاج" ويحسبها. هدفك ${n(t.protein)} غ بروتين اليوم.`,
        };

  const nightBarrier = profile.barriers.includes("night");
  const restore: Mission =
    energy === 1
      ? {
          id: "restore",
          kind: "restore",
          title: "نوم أبكر بنص ساعة",
          detail: "قلة النوم ترفع هرمون الجوع وتضعف الإرادة. الليلة استثمار.",
        }
      : nightBarrier
        ? {
            id: "restore",
            kind: "restore",
            title: "المطبخ يسكّر الساعة ٩",
            detail: "بعدها شاي أو ماء فقط. خطتك لو جاك جوع الليل جاهزة في صفحة المدرب.",
          }
        : {
            id: "restore",
            kind: "restore",
            title: "٨ أكواب ماء",
            detail: "ابدأ بكوب قبل كل وجبة؛ يساعد على الشبع ويقلل الخلط بين العطش والجوع.",
          };

  return [move, eat, restore];
}

export const ENERGY_COPY: Record<Energy, { label: string; line: string }> = {
  1: { label: "طاقتي تحت", line: "عادي. اليوم نحافظ على الخيط بأصغر خطوة ممكنة." },
  2: { label: "نص نص", line: "يوم متوازن: خطوات ثابتة بدون ضغط." },
  3: { label: "فل طاقة", line: "استغلها! اليوم نبني عضل ونسبق الخطة." },
};

File: android/core/src/main/kotlin/app/sanad/core/Science.kt
package app.sanad.core

import java.time.LocalDate
import java.time.temporal.ChronoUnit
import kotlin.math.max
import kotlin.math.min
import kotlin.math.pow
import kotlin.math.roundToInt

/** طاقة تقريبية لكل كغ من نسيج الجسم المفقود (خليط دهون وعضل). */
const val KCAL_PER_KG = 7700.0

fun calorieFloor(sex: Sex): Int = if (sex == Sex.F) 1200 else 1500

private fun roundTo(x: Double, step: Int): Int = ((x / step).roundToInt()) * step

fun bmi(weightKg: Double, heightCm: Double): Double {
    val h = heightCm / 100
    return weightKg / (h * h)
}

/** معادلة Mifflin-St Jeor — الأدق بين المعادلات الشائعة للبالغين. */
fun bmr(sex: Sex, weightKg: Double, heightCm: Double, age: Int): Double {
    val base = 10 * weightKg + 6.25 * heightCm - 5 * age
    return if (sex == Sex.M) base + 5 else base - 161
}

fun formulaTdee(p: Profile, weightKg: Double): Double = bmr(p.sex, weightKg, p.heightCm, p.age) * p.activity.factor

/** وزن مرجعي للبروتين: عند زيادة الوزن = الوزن عند BMI 25 + ربع الفائض. */
fun referenceWeight(weightKg: Double, heightCm: Double): Double {
    val h = heightCm / 100
    val at25 = 25 * h * h
    return if (weightKg <= at25) weightKg else at25 + 0.25 * (weightKg - at25)
}

/** ١٫٢–١٫٦ غ/كغ أثناء النزول للحفاظ على العضل؛ نستهدف ١٫٥ من الوزن المرجعي. */
fun proteinTarget(weightKg: Double, heightCm: Double): Int = roundTo(1.5 * referenceWeight(weightKg, heightCm), 5)

data class Targets(
    val tdee: Int,
    val kcal: Int,
    val protein: Int,
    val weeklyLossKg: Double,
    val steps: Int,
    val water: Int,
    val floorApplied: Boolean,
    val adaptive: AdaptiveResult? = null,
)

fun computeTargets(p: Profile, weightKg: Double, adaptive: AdaptiveResult? = null): Targets {
    val tdee = adaptive?.tdee?.toDouble() ?: formulaTdee(p, weightKg)
    val loss = p.pace.weeklyRate * weightKg
    val raw = tdee - loss * KCAL_PER_KG / 7
    val floor = calorieFloor(p.sex)
    return Targets(
        tdee = roundTo(tdee, 10),
        kcal = roundTo(max(raw, floor.toDouble()), 10),
        protein = proteinTarget(weightKg, p.heightCm),
        weeklyLossKg = (loss * 100).roundToInt() / 100.0,
        steps = p.activity.baseSteps,
        water = 8,
        floorApplied = raw < floor,
        adaptive = adaptive,
    )
}

/* ---------------- الوزن الاتجاهي ---------------- */

data class TrendPoint(val date: String, val kg: Double, val trend: Double)

fun daysBetween(a: String, b: String): Long = ChronoUnit.DAYS.between(LocalDate.parse(a), LocalDate.parse(b))

/** EMA بمعامل ٠٫١ يومياً؛ الفجوة الأطول تعطي القراءة الجديدة وزناً أكبر. */
fun trendWeights(points: List<Pair<String, Double>>, alpha: Double = 0.1): List<TrendPoint> {
    var trend: Double? = null
    var last: String? = null
    return points.sortedBy { it.first }.map { (date, kg) ->
        val t = trend
        trend = if (t == null || last == null) kg else {
            val gap = max(1L, daysBetween(last!!, date)).toInt()
            val a = 1 - (1 - alpha).pow(gap)
            t + a * (kg - t)
        }
        last = date
        TrendPoint(date, kg, (trend!! * 100).roundToInt() / 100.0)
    }
}

fun weightPoints(days: Collection<DayLog>): List<Pair<String, Double>> =
    days.mapNotNull { d -> d.weightKg?.let { d.date to it } }

/* ---------------- الحرق التكيّفي ---------------- */

data class AdaptiveResult(val tdee: Int, val confidence: Double, val loggedDays: Int, val windowDays: Long)

/**
 * نافذة حتى ٢١ يوماً: الحرق ≈ متوسط الأكل − (تغيّر الوزن الاتجاهي × ٧٧٠٠ ÷ الأيام)،
 * ممزوج مع المعادلة حسب كمية البيانات، ومحدود بـ ±٢٥٪.
 */
fun adaptiveTdee(p: Profile, days: Collection<DayLog>, today: String): AdaptiveResult? {
    val window = days.filter { daysBetween(it.date, today) in 0..20 }.sortedBy { it.date }
    val intakeDays = window.filter { it.intake >= 600 }
    val weights = weightPoints(window)
    if (intakeDays.size < 7 || weights.size < 3) return null
    val trend = trendWeights(weights)
    val span = daysBetween(trend.first().date, trend.last().date)
    if (span < 7) return null
    val avgIntake = intakeDays.map { it.intake }.average()
    val measured = avgIntake - (trend.last().trend - trend.first().trend) * KCAL_PER_KG / span
    val formula = formulaTdee(p, trend.last().trend)
    val confidence = min(1.0, (intakeDays.size / 14.0) * min(1.0, span / 14.0))
    val blended = formula + (measured - formula) * confidence
    val clamped = min(formula * 1.25, max(formula * 0.75, blended))
    return AdaptiveResult(roundTo(clamped, 10), (confidence * 100).roundToInt() / 100.0, intakeDays.size, span)
}

/* ---------------- سلسلة الاستمرار: لا تفوّت مرتين ---------------- */

fun isCounted(d: DayLog?): Boolean =
    d != null && (d.done.isNotEmpty() || d.meals.isNotEmpty() || d.workouts.isNotEmpty() || d.weightKg != null || d.lapses.isNotEmpty())

data class ThreadState(val length: Int, val rescueToday: Boolean, val best: Int)

fun addDays(date: String, n: Long): String = LocalDate.parse(date).plusDays(n).toString()

fun computeThread(days: Map<String, DayLog>, today: String): ThreadState {
    val first = days.keys.minOrNull() ?: return ThreadState(0, false, 0)
    var best = 0
    var run = 0
    var misses = 0
    var cursor = first
    while (cursor <= today) {
        if (isCounted(days[cursor])) {
            run++; misses = 0
        } else if (cursor != today) {
            misses++
            if (misses >= 2) run = 0
        }
        best = max(best, run)
        cursor = addDays(cursor, 1)
    }
    val yesterday = addDays(today, -1)
    val rescue = run > 0 && !isCounted(days[yesterday]) && !isCounted(days[today]) && first <= yesterday
    return ThreadState(run, rescue, best)
}

/** حالة كل يوم في نسيج السدو (من الأقدم للأحدث). */
enum class WeaveCell { WOVEN, HELD, BROKEN, TODAY, BEFORE }

fun weaveCells(days: Map<String, DayLog>, today: String, count: Int, firstDay: String?): List<WeaveCell> {
    val dates = (0 until count).map { addDays(today, (it - count + 1).toLong()) }
    val counted = dates.map { isCounted(days[it]) }
    return dates.mapIndexed { i, d ->
        when {
            firstDay != null && d < firstDay -> WeaveCell.BEFORE
            counted[i] -> WeaveCell.WOVEN
            d == today -> WeaveCell.TODAY
            else -> {
                val prevMissed = i > 0 && !counted[i - 1] && !(firstDay != null && dates[i - 1] < firstDay)
                val nextMissed = i < count - 1 && !counted[i + 1] && dates[i + 1] != today
                if (prevMissed || nextMissed) WeaveCell.BROKEN else WeaveCell.HELD
            }
        }
    }
}

/* ---------------- خطة اليوم ---------------- */

enum class MissionKind { MOVE, EAT, RESTORE }

data class Mission(val id: String, val kind: MissionKind, val title: String, val detail: String, val routineId: String? = null)

fun dayMissions(energy: Energy, time: TimeBudget, t: Targets, p: Profile): List<Mission> {
    if (p.ramadan) return ramadanMissions(energy, time, t)
    val proteinMeal = roundTo(t.protein / 3.0, 5)
    val move = when {
        energy == Energy.LOW || time == TimeBudget.TWO -> Mission(
            "move", MissionKind.MOVE, "دقيقتين حركة فقط",
            "٣ تمارين هادئة وأنت بمكانك. الهدف تحافظ على السلسلة، مو تتعب.",
            if (energy == Energy.LOW) "reset-2" else "wake-2",
        )
        time == TimeBudget.TEN -> Mission(
            "move", MissionKind.MOVE,
            if (energy == Energy.HIGH) "١٠ دقائق قوة للجسم كامل" else "١٠ دقائق حركة بدون قفز",
            "تمارين بوزن الجسم تحمي عضلاتك وأنت تنزل وزن.",
            if (energy == Energy.HIGH) "strength-10" else "low-impact-10",
        )
        else -> Mission(
            "move", MissionKind.MOVE,
            if (energy == Energy.HIGH) "٢٠ دقيقة قوة + مشي" else "٢٠ دقيقة مشي خفيف",
            if (energy == Energy.HIGH) "جلسة قوة كاملة؛ أهم استثمار للحفاظ على العضل." else "امشِ بإيقاع مريح. هدف خطواتك ${ar(t.steps)}.",
            if (energy == Energy.HIGH) "strength-20" else "walk-20",
        )
    }
    val eat = if (energy == Energy.LOW) Mission(
        "eat", MissionKind.EAT, "ابدأ وجبتك الجاية بالبروتين",
        "حوالي ${ar(proteinMeal)} غ (بيض، زبادي، دجاج، تونة). التعب يرفع الجوع — البروتين يهدّيه.",
    ) else Mission(
        "eat", MissionKind.EAT, "سجّل وجباتك — ولو بجملة",
        "قل لسند \"تغديت كبسة دجاج\" ويحسبها. هدفك ${ar(t.protein)} غ بروتين اليوم.",
    )
    val restore = when {
        energy == Energy.LOW -> Mission("restore", MissionKind.RESTORE, "نوم أبكر بنص ساعة", "قلة النوم ترفع هرمون الجوع وتضعف الإرادة. الليلة استثمار.")
        Barrier.NIGHT in p.barriers -> Mission("restore", MissionKind.RESTORE, "المطبخ يسكّر الساعة ٩", "بعدها شاي أو ماء فقط. خطتك لو جاك جوع الليل جاهزة عند المدرب.")
        else -> Mission("restore", MissionKind.RESTORE, "٨ أكواب ماء", "ابدأ بكوب قبل كل وجبة؛ يساعد على الشبع.")
    }
    return listOf(move, eat, restore)
}

val ENERGY_COPY = mapOf(
    Energy.LOW to ("طاقتي تحت" to "عادي. اليوم نحافظ على السلسلة بأصغر خطوة ممكنة."),
    Energy.MID to ("نص نص" to "يوم متوازن: خطوات ثابتة بدون ضغط."),
    Energy.HIGH to ("فل طاقة" to "استغلها! اليوم نبني عضل ونسبق الخطة."),
)
